Adds new `bit` element_type for dense_vectors

benwtrent opened the pull request:
This commit adds `bit` vector support by adding `element_type: bit` for vectors. This new element type works for indexed and non-indexed vectors. Additionally, it works with `hnsw` and `flat` index types. No quantization based codec works with this element type, this is consistent with `byte` vectors.

`bit` vectors accept up to `32768` dimensions in size and expect vectors that are being indexed to be encoded either as a hexidecimal string or a `byte[]` array where each element of the `byte` array represents `8` bits of the vector.

`bit` vectors support script usage and regular query usage. When indexed, all comparisons done are `xor` and `popcount` summations (aka, hamming distance), and the scores are transformed and normalized given the vector dimensions. Note, indexed bit vectors require `l2_norm` to be the similarity.

For scripts, `l1norm` is the same as `hamming` distance and `l2norm` is `sqrt(l1norm)`. `dotProduct` and `cosineSimilarity` are not supported.

Note, the dimensions expected by this element_type are always to be divisible by `8`, and the `byte[]` vectors provided for index must be have size `dim/8` size, where each byte element represents `8` bits of the vectors.

closes: https://github.com/elastic/elasticsearch/issues/48322

benwtrent committed:
Add binary vector support for HNSW and FLAT indices

benwtrent committed:
iter

benwtrent committed:
iter

benwtrent committed:
iter

benwtrent committed:
iter

benwtrent committed:
Merge remote-tracking branch 'upstream/main' into feature/binary-vector-support

benwtrent committed:
iter

benwtrent committed:
iter

benwtrent committed:
iter

benwtrent committed:
removed unnecessary change

benwtrent committed:
iter

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @benwtrent, I've created a changelog YAML for you. *Note* that since this PR is labelled `release highlight`, you need to update the changelog YAML to fill out the extended information sections.

benwtrent committed:
Update docs/changelog/110059.yaml

benwtrent commented:
I am planning on adding docs soonish

benwtrent committed:
Merge remote-tracking branch 'upstream/main' into feature/binary-vector-support

benwtrent committed:
fixing test

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/codec/vectors/ES815BitFlatVectorsFormat.java:
```diff
@@ -0,0 +1,139 @@
+                    case DOT_PRODUCT, MAXIMUM_INNER_PRODUCT, COSINE, EUCLIDEAN -> new HammingScorerSupplier(randomAccessVectorValuesBytes);
+                };
+            }
+            throw new IllegalArgumentException("Unsupported vector type");
```
should we also add "Unsupported vector type or similarity function", or we don't expect changes to `VectorSimilarityFunction`?

benwtrent replied:
I can add that to the error, but if the function does ever get updated, we would fail to compile, so we would likely have to add some handling for it anyways.

benwtrent committed:
adding docs, fixing test

benwtrent commented on modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/146_dense_vector_bit_basic.yml:
```diff
@@ -0,0 +1,399 @@
+        index: test-index
+        body:
+          settings:
+            number_of_replicas: 0
```
@mayya-sharipova good catch, I will adjust.

benwtrent committed:
fixing tests

benwtrent committed:
fixing test

benwtrent committed:
fixing moar tests

mayya-sharipova commented on docs/reference/mapping/types/dense-vector.asciidoc:
```diff
@@ -183,11 +183,23 @@ The following mapping parameters are accepted:
+indexes a single bit per dimension. Useful for very high-dimensional vectors or models that specifically support bit vectors.
+NOTE: when using `bit`, the number of dimensions must be a multiple of 8 and must represent the number of bits.
+
+====
```
Should we also update `[[dense-vector-similarity]]` section to explain that all these functions are not relevant for bit vectors and their scores are computed based on hamming distance?

benwtrent replied:
@mayya-sharipova yeah, I can do that

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:
```diff
@@ -795,6 +1045,7 @@ public enum VectorSimilarity {
             float score(float similarity, ElementType elementType, int dim) {
                 return switch (elementType) {
                     case BYTE, FLOAT -> 1f / (1f + similarity * similarity);
+                    case BIT -> dim - (similarity * dim);
```
I am confused about this. What does`similarity` represent here?  Is it the number of similar bits?  Shouldn't than the score be: ` similarity/dim`?

benwtrent replied:
The hamming score transformation is:
```
((a.length * Byte.SIZE) - VectorUtil.xorBitCount(a, b)) / (float) (a.length * Byte.SIZE);
```

`a.length * Byte.SIZE` is just the `dim` here as in the mapping we account for it representing bits. In Lucene, we account for the `byte` values (thus we need to correct it). 

As for if this is the correct way to transform the score. Let's take it case by case. Really we only have two cases, `dotProduct` (which covers cosine, max-inner-product), and `euclidean`. 

For `euclidean`, the `l2norm` distance is just `hamming`. But with `l2norm` directly, values that are higher are further away. So we must do a scoring inversion. So, doing `bitDimensionCount - xor` gives us the number of bits that are the same (inverting the l2norm so that bigger values give us a smaller score, e.g. vectors further away score lower). I then normalize by `dim` to scale us similarly with all our other vector scores (the max possible _score for bit vectors is `dim` as we just count the number of bits that are the same).


For dotProduct, surprisingly, the same transformation works. But it has the assumption that the provided vectors are bit-quantized (meaning transformed to `0` and `1` around some median, and thus having the exact same bits in that case is beneficial).


Maybe if we have `l2norm` be the inverted `xor` score, we should consider making `dotProduct` an actual `dotProduct` (a `bit AND` score). I did this originally. 

@tveasey @jimczi what do you think? How should we handle dotProduct (and its family of max inner product and cosine)?

tveasey replied:
As discussed offline there are subtleties around this that I'm just noting here for completeness.

The crux of the matter is how the vectors are quantised. Assuming you want minimum mean square error estimates and you quantise as x_i <= median_i |-> 0 and x_i > median |-> 1 then one can work through minimum mean squared expected estimators for each type of distance / similarity function. Generally, I think people are assuming each component is IID (which usually isn't too bad for learned representations). This allows you to compute distances/similarities directly on bit vectors up to some constant scaling.

Taking L2 as a prototype we have

dist(0, 0) = (E[x | x <= median] - E[x | x <= median])^2 = 0 (similar for dist(1, 1))
dist(0, 1) = (E[x | x > median] - E[x | x <= median])^2 = k^2

so our best estimate for L2  would be k^2 * count(x ^ y). (One can run a similar sort of analysis for dot.)

mayya-sharipova replied:
@benwtrent @tveasey Thanks for detailed explanations.

@benwtrent If similarity in this funciton is ` hamming`, why `score` is not calculated as you defined in documentation: `(numBits - hamming(a, b)) / numBits` , so will be 

`(dim - similarity) / dim` ?

benwtrent replied:
@mayya-sharipova you are correct, let me fix it!

benwtrent committed:
restrict to l2_norm for bit vectors

benwtrent commented:
@mayya-sharipova I just pushed a commit that:

 - clarified the similarity section with bit vectors
 - Defaults to l2_norm when vectors are indexed with bit element type
 - Disallows other similarity functions when the element type is bit & indexed.

benwtrent committed:
fixing test

benwtrent committed:
fixing similarity -> score transformation

benwtrent committed:
fixing tests

mayya-sharipova commented on server/src/test/java/org/elasticsearch/index/codec/vectors/BaseKnnBitVectorsFormatTestCase.java:
```diff
@@ -0,0 +1,149 @@
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+
+abstract class BaseKnnBitVectorsFormatTestCase extends BaseIndexFileFormatTestCase {
```
What we are testing in this file related to this PR?

mayya-sharipova reviewed (commented):
@benwtrent  Great work!  +1 for extensive testing and documenation.

I have only 1 topic left to discuss about script functions for bit vectors:
- Does it make sense to have `cosine`, `dotProduct` and `l1norm` script functions for bit vectors?  If yes, do they use the same formula as byte vectors?
- Please update `vector-functions.asciidoc` to add that `hamming` distance function is also applicable for bit vectors.

ChrisHegarty commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:
```diff
@@ -665,13 +694,238 @@ ByteBuffer createByteBuffer(IndexVersion indexVersion, int numBytes) {
+                        yield index * Byte.SIZE;
+                    }
+                    case VALUE_STRING -> {
+                        byte[] decodedVector = HexFormat.of().parseHex(context.parser().text());
```
It might be fine, but what exception/error will be get if this parsing fails?

benwtrent replied:
`failed to parse field...`

We get an invalid format error and it gets translated to a bad request.

benwtrent commented:
>  Does it make sense to have cosine, dotProduct and l1norm script functions for bit vectors? If yes, do they use the same formula as byte vectors?

It does make sense to me, and no they do not use the same formula as byte vectors. I need to update the docs there.

> Please update vector-functions.asciidoc to add that hamming distance function is also applicable for bit vectors.

Yep, I will do that asap!

benwtrent commented:
@mayya-sharipova actually, thinking about the complexities of what a bit dotProduct means, I am going to remove support for it directly from script functions.

We can revisit this in a future PR.

benwtrent committed:
adjusting script support and adding tests

benwtrent commented:
@mayya-sharipova just pushed:

 - I dropped support for cosine & dot product in the script
 - Adding docs about this and what l1, l2, and hamming are for bit vectors
 - Also, I noticed I didn't have tests & such for magnitude & bit vectors, added that and optimized the magnitude calculation.

benwtrent committed:
fixing tests

mayya-sharipova commented on docs/reference/vectors/vector-functions.asciidoc:
```diff
@@ -17,6 +16,8 @@ This is the list of available vector functions and vector access methods:
 6. <<vector-functions-accessing-vectors,`doc[<field>].vectorValue`>> – returns a vector's value as an array of floats
 7. <<vector-functions-accessing-vectors,`doc[<field>].magnitude`>> – returns a vector's magnitude
 
+NOTE: The `cosineSimilarity` and `dotProduct` functions are not supported for `bit` vectors.
```
Thanks Ben!

Also  Line 196 says: The `hamming` function calculates {wikipedia}/Hamming_distance[Hamming distance] between a given query vector and document vectors. It is only **available for byte vectors.**.

We should correct this.

benwtrent committed:
fixing docs

benwtrent committed:
Update 110059.yaml

benwtrent commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into feature/binary-vector-support

mayya-sharipova commented:
@benwtrent Thanks, the code LGTM.

One thing that concerns me how we use `l2_norm` similarity parameter for defining bit vectors:
```
"bit_vector": {
    "type": "dense_vector",
    "dims" : 40,
    "element_type" : "bit",
    "similarity": "l2_norm"
}
```

because we don't actually use `l2_norm`, we use `l1_norm` or `hamming` for scoring calculations, but we don't have those similarity metrics defined in mapping.  
Could it be confusing for users that we define "l2_norm", but never use  in score calculations? What do you think>

One suggestion I have is NOT allow to provide any "similarity" metric for bit vectors in mappings for users, and say in documentation that "similarity" is not defined for bit vectors. What do you think of this suggestion?

benwtrent commented:
@mayya-sharipova 

> One thing that concerns me how we use `l2_norm` similarity parameter for defining bit vectors:
> ...
> because we don't actually use l2_norm, we use l1_norm or hamming for scoring calculations, but we don't have those similarity metrics defined in mapping.

So, our `l2_norm` calculation for all our other data types is just the `squareDifference`, we don't actually take the square root.

Lets take some `bits` for example: `1011` and `1001`. 

`l2_norm` would be `(1-1)^2 + (0-0)^2 + (1-0)^2 + (1-1)^2` = `0 + 0 + 1 + 0` = `1`
`hamming` is then `1^1 + 0^0 + 1^0 + 1^1` = `0 + 0 + 1 + 0` = `1`

For `bit` vectors & how we already score things, `l1_norm` `l2_norm` and `hamming` are all the same.

benwtrent commented:
@mayya-sharipova also, if users don't provide any similarity, we default to `l2_norm` for `bit`. So, its not strictly necessary at all.

mayya-sharipova commented:
@benwtrent Thanks for the detailed explanation. I guess it is ok to keep `l2_norm` as defafult.

benwtrent commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into feature/binary-vector-support
